From accceb395ed45ec7224fa5e36388e59548a64c16 Mon Sep 17 00:00:00 2001
From: Richard Liu <xuegang.liu@nxp.com>
Date: Fri, 12 Oct 2018 00:56:56 +0800
Subject: [PATCH 4814/5242] MA-12874 [#imx-1203] Fix random kernel panic in
 viv_gem_prime_import

commit  340286d57bf7cec097b2b539a0e74a447b7c9339 from
https://source.codeaurora.org/external/imx/linux-imx.git

Fix random kernel panic in viv_gem_prime_import when run
android CTS or monkey test. The panic reason is dmabuf size
no meet page alignment request. Do buffer size page alignment
when exported dmabuf from vidmem to fix the issue.

Test case:
while true;do monkey --pct-syskeys 0 -v 100000;done

Kernel panic stack:
[ 140.322048] PC is at drm_gem_private_object_init+0x58/0x5c
[ 140.327532] LR is at viv_gem_prime_import+0xdc/0x11c
[ 140.332497] pc : [<ffff000008718a00>] lr : [<ffff000008b7d6f0>] pstate: 40000145
[ 140.339889] sp : ffff00000d45ba40
[ 140.343199] x29: ffff00000d45ba40 x28: 000000000000000c
[ 140.348507] x27: ffff00000d45bd58 x26: ffff8008ed08e200
[ 140.353815] x25: ffff8008f6eab5b8 x24: 000000000000000c
[ 140.359123] x23: ffff00000d45bd58 x22: ffff8008a34a4000
[ 140.364431] x21: ffff8008f6eab000 x20: ffff000009c1a000
[ 140.369739] x19: ffff80083d7ffa00 x18: 0000f1d2ac2ac8aa
[ 140.375047] x17: 0000f1d2c70690e0 x16: ffff0000082add48
[ 140.380356] x15: 0000000000000000 x14: 0000000000000000
[ 140.385663] x13: 0000000000000001 x12: 00000000ffffffff
[ 140.390971] x11: 0000f1d2ac2ac4f0 x10: 0000f1d2ac2ac4a8
[ 140.396280] x9 : 0000000000000000 x8 : ffff80083d7ffb00
[ 140.401587] x7 : 0000000000000000 x6 : 000000000000003f
[ 140.406896] x5 : 0000000000000040 x4 : 0000000000000000
[ 140.412204] x3 : 0000000000000040 x2 : 0000000000002040
[ 140.417512] x1 : ffff80083d7ffa00 x0 : ffff8008f6eab000

Change-Id: I787b0bcfd4e5e79996eda073fd1a12924c97d170
Signed-off-by: Richard Liu <xuegang.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../hal/kernel/gc_hal_kernel_video_memory.c        |    1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_video_memory.c b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_video_memory.c
index c1b6629a..2eb6a65 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_video_memory.c
+++ b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_video_memory.c
@@ -2946,6 +2946,7 @@ static void _dmabuf_kunmap(struct dma_buf *dmabuf, unsigned long offset, void *p
         {
             physical = node->VidMem.memory->physical;
             bytes = node->VidMem.bytes;
+            bytes = bytes & (~(PAGE_SIZE -1));
         }
         else
         {
-- 
1.7.9.5

